---
title: "R Notebook"
output: html_notebook
author: Ignaitus Pang ` i.pang at unsw.edu.au '
---


Be kind to your future self, comment your codes. - Paraphrased from Sue Wilson

```{r}
if( !require(pacman)) {
  install.packages("pacman")
  library(pacman)
}

p_load(tidyverse)
p_load(here)
p_load(readxl)
p_load(vroom)
# p_load(multidplyr)
 
base_dir <- here::here()

source( file.path( base_dir, "Source/Common/common_parameters.R") )
source( file.path( source_dir, "Common/helper_functions.R") )

```

## Global Parameters
```{r}
# cluster <- new_cluster(8)
# cluster_library( cluster, packages = c("tidyverse", "here"))

list_of_e_values_to_test <-  c(  10^-6, 10^-5, 10^-4, 10^-3, 10^-2, 10^-1, 1 ) 
best_e_value_to_use <- 10^-3
```

nohup time Rscript --vanilla -e "rmarkdown::render('parse_blast_sepsis_strain_to_uniprot.Rmd', output_file='parse_blast_sepsis_strain_to_uniprot.html') " > parse_blast_sepsis_strain_to_uniprot.log 2>&1 &


## Directories Management

### Create output directories for storing results from applying a range of different BlastP E-value thresholds

```{r}

purrr::map( list_of_e_values_to_test, function(my_e_value_cutoff ) {
  kegg_annotations_dir <- file.path( oc_results_dir, "KEGG_Annotations", paste( "E_value_", my_e_value_cutoff, sep=""), "Map_to_Uniprot" )
  create_dir_if_not_exists(kegg_annotations_dir)
} ) 

```

### File names for the KEGG database data
```{r}
kegg_dir <- file.path( owncloud_dir, "Annotations_and_Mapping/KEGG_data_files/genes/links")

kegg_genes_to_uniprot_cleaned_file <- file.path( kegg_dir, "genes_uniprot.list") # KEGG gene ID to UniProt accession table
kegg_genes_to_pathway_cleaned_file <- file.path( kegg_dir, "genes_pathway.list") # Tablve for converting KEGG gene ID to KEGG pathway ID 
kegg_genes_to_ko_file <- file.path( kegg_dir, "genes_ko.list") # Table for converting KEGG gene ID to KEGG ortholog (ko) group ID 
```

### System glob path for finding all the BlastP outputs
```{r}
blast_to_uniprot_file_path <- file.path( oc_results_dir, "Blast_Strain_To_Uniprot_DB/TSV")
```

### Save all of the BlastP results into one file for easy management later 
```{r}
create_dir_if_not_exists(file.path( oc_results_dir,
                                           "Blast_Strain_To_Uniprot_DB/Merged_TSV/") )

blast_to_uniprot_merged_file <- file.path( oc_results_dir,
                                           "Blast_Strain_To_Uniprot_DB/Merged_TSV/blast_to_uniprot_merged.tsv")

```


## Read Genome statistics
I thought I will be a bit more thorough in listing the columns descriptions (more than usual) for this section because it could be hard to understand without these information or opening the Excel file containing the information. 
```{r}

## The Excel file containing a number of tables with information about the Genomes
genome_statistics_file <- file.path( data_dir, "Genomes_List/num_seq_per_strain_and_accession.xlsx") 

# Table: num_seq_per_accession 
# Description: The number of protein sequences per chromosome / plasmid
# Column 1: chromosome_refseq_id, character, e.g. NZ_LR129840.1
# Column 2: num_sequences, integer, e.g. 2075
num_seq_per_accession  <- read_xlsx(path=genome_statistics_file, sheet="num_seq_per_accession")

# table: strain_to_accessions
# Description: Get information about the Species and Strain name as compared to the Chromosome/Plasmid RefSeq ID
# Column 1: Species, text, e.g. Streptococcus pyogenes	
# Column 2: Strain, text, e.g. HKU419
# Column 3: refseq_ids, text, e.g. NZ_LR595848.1; NZ_LR595849.1	
#  (The chormosome RefSeq ID. Note that there could be multiple RefSeq ID per line separate by semi-colons and spaces.
#   Some cleaning is done below to clean this up).
# Column 4: Species_strain_file_name, e.g. Streptococcus.pyogenes.HKU419.faa	
# Column 5: List_of_files_to_concat, e.g. NZ_LR595848.1.faa NZ_LR595849.1.faa 
#           (The list of files to concatenate together, including the files containing the chromosomes and plasmids)
strain_to_accessions   <- read_xlsx(path=genome_statistics_file, sheet="strain_to_accessions")

# Table: chromosome_id_to_protein_id
# Description: Get all the Protein RefSeq ID associated with each Chromosome / Plasmid RefSeq ID
# Column 1: chromosome_refseq_id
# Column 2: protein_refseq_id
# Column 3: protein_description
chromosome_id_to_protein_id <- read_xlsx(path=genome_statistics_file, sheet="chromosome_id_to_protein_id")

# Table: strain_to_accessions_cleaned
# Description: Convert Species and Strain name to chromosome/plasmid RefSeq ID. 
#   Clean the refseq_id column using the 'separate_rows' fucntion such that there 
#   is only one refseq_id per row. 
# Column 1: Species, text, e.g. Streptococcus pyogenes	
# Column 2: Strain, text, e.g. HKU419
# Column 1: chromosome_refseq_id, character, e.g. NZ_LR129840.1
strain_to_accessions_cleaned <- strain_to_accessions %>%
                                dplyr::select( Species, Strain, refseq_ids) %>%
                                separate_rows( refseq_ids, sep = "; ") %>%
                                dplyr::rename( chromosome_refseq_id = "refseq_ids")

# table: num_seq_per_accession 
# Description: The number of protein sequences per chromosome / plasmid
# Column 1: chromosome_refseq_id, character, e.g. NZ_LR129840.1
# Column 2: num_sequences, integer, e.g. 2075
num_seq_per_accession_cleaned <- strain_to_accessions_cleaned %>% 
                                 left_join( num_seq_per_accession, by=c("chromosome_refseq_id" = "chromosome_refseq_id") ) 

```


### Read BlastP outputs and merge the BlastP ouptut tables together
```{r}
column_names <- c("query_acc.ver", "subject_acc.ver", "perc_identity", 
                  "alignment_length", "mismatches", "gap_opens", 
                  "q_start", "q_end", "s_start", "s_end", 
                  "evalue", "bit_score")

list_of_files <- fs::dir_ls(path = blast_to_uniprot_file_path,
                            glob = "*")

## Read in the files using vroom
temp_table <- vroom(list_of_files, id="Comparison", col_names =column_names, comment="#", delim ="\t" )  

cleaned_table <- temp_table %>%
                ## Remove all Streptococcus pneumoniae data
                ## Clean the file names so that '-' is only used to separate the Species name and the 
                ## strain name within the file name. 
                mutate( Comparison =  str_replace_all( Comparison, ".*/(.*).tab", "\\1")  %>%
                         str_replace_all( "03-311-0071", "03/311/0071"  ) %>%
                         str_replace_all( "180.2", "180/2") %>%
                         str_replace_all( "180.15", "180/15")        ) %>%
                ## Filter the Uniprot accessions, only take the input file if the Subject Database is a Uniprot Entry
                filter( str_detect( subject_acc.ver, "^(sp|tr)\\|")) %>%
                ## Separate the Species and the Strain from the file names
                separate(col="Comparison",  sep ="[-\\.]", into=c("Query_a", "Query_b", "Query_Strain")) %>%
                mutate( Query_Species = paste(Query_a,  Query_b))  %>%
                dplyr::select ( -Query_a, -Query_b ) %>%
                ## Clean up the strain names
                mutate( Query_Strain = str_replace( Query_Strain,  "03/311/0071", "03-311-0071") %>%
                                       str_replace_all( "180/2", "180.2") %>%
                                       str_replace_all( "180/15", "180.15") ) %>%
                dplyr::select( one_of(c( "Query_Species", "Query_Strain",  column_names)))  %>%
                filter( Query_Species != "Streptococcus pneumoniae" )


rm(temp_table)


```



## Clean up the BlastP output table and rank the hits by e-values
```{r}

## Rank each query to subject blast searches by the E-value score.
ranked_table <- cleaned_table  %>%
  group_by( query_acc.ver  ) %>%
  nest() %>%
  #partition(query_acc.ver, .cluster = cluster) %>%
  mutate( data = purrr::map( data, function(x) { x %>% mutate( rank = row_number(evalue)) }  )) %>%
  #collect() %>%
  unnest(data) %>%
  ungroup()

hits_to_uniprot_table <- ranked_table  %>%
    ## Extract the Uniprot Accession number of the hit
   mutate( subject_uniprot_acc = ifelse( str_detect(subject_acc.ver, "|"), 
                                         str_replace( subject_acc.ver, "(sp|tr)\\|(.*)\\|.*", "\\2"), 
                                         NA) ) %>%
  # Clean up species and strain for Klebsiella pneumoniae versus Klebsiella variicola"
  mutate ( Query_Species = case_when( Query_Species == "Klebsiella pneumoniae"  &
                                      Query_Strain %in% c( "AJ055", 
                                                           "AJ292", 
                                                           "03-311-0071", 
                                                           "04153260899A") ~ "Klebsiella variicola",
                                      TRUE ~ Query_Species))


# hits_to_uniprot_table %>%
#   distinct( Query_Species, Query_Strain)

write_tsv( hits_to_uniprot_table, path=blast_to_uniprot_merged_file) 

```

## Summary statistics of how many proteins have UniProt accession
```{r}
summary_statistics_hits_to_uniprot <-  function( hits_to_uniprot_table, my_e_value_cutoff ) {
    hits_to_uniprot_summary_statistics <- hits_to_uniprot_table %>%
      filter( evalue < my_e_value_cutoff  ) %>%
      inner_join( chromosome_id_to_protein_id, by=c("query_acc.ver" = "protein_refseq_id") ) %>%
      inner_join ( strain_to_accessions_cleaned, by =c("chromosome_refseq_id" = "chromosome_refseq_id", 
                                                       "Query_Species" = "Species",
                                                       "Query_Strain" = "Strain")) %>%
      distinct(Query_Species, Query_Strain, chromosome_refseq_id, query_acc.ver ) %>%
      group_by(Query_Species, Query_Strain, chromosome_refseq_id) %>%
      # partition(Query_Species, Query_Strain, chromosome_refseq_id, .cluster = cluster) %>%
      summarise( num_proteins_in_pathways = n()) %>%
      # collect() %>%
      inner_join(  num_seq_per_accession, by=c("chromosome_refseq_id" = "chromosome_refseq_id")   ) %>%
      dplyr::rename( total_num_proteins_in_proteome = "num_sequences") %>%
      distinct() 
    
    return( hits_to_uniprot_summary_statistics)
    
}


summary_statistics_hits_to_uniprot_save_file <- function(hits_to_uniprot_table, my_e_value_cutoff ) { 
  
    kegg_annotations_dir <- file.path( oc_results_dir, "KEGG_Annotations",
                                       paste( "E_value_", my_e_value_cutoff, sep=""), "Map_to_Uniprot" )

    resutls_table <- summary_statistics_hits_to_uniprot( hits_to_uniprot_table, my_e_value_cutoff )
    
    write_tsv(resutls_table,
              path=file.path(kegg_annotations_dir, 
                             "mapping_uniprot_hits_summary_statistics.tsv") ) 
}

```


# Convert from BlastP Hits to KEGG pathways
```{r}

convert_hits_to_pathways <- function( hits_to_uniprot_table, my_e_value_cutoff ) {
  hits_to_pathways_table_cleaned <- hits_to_uniprot_table %>%
      filter( evalue < my_e_value_cutoff  )  %>%
      filter( str_detect( subject_acc.ver, "^(sp|tr)\\|")) %>%
      inner_join( kegg_genes_to_uniprot_cleaned, by=c("subject_uniprot_acc" = "uniprot_acc_cleaned")) %>%
      inner_join( kegg_genes_to_pathway_cleaned, by=c("kegg_entrez_id" = "kegg_entrez_id"))  %>%
      filter( !is.na(kegg_pathway_id)) %>%
      mutate( cleaned_kegg_pathway_id = str_replace( kegg_pathway_id, ".*(\\d{5}$)", "\\1")   ) %>%
      ## Map protein hits to KEGG ortholog groups 
      ## These KEGG ortholog groups represents proteins in the KEGG pathways. Duplicated proteins will have the same KEGG ortholog group label. 
      inner_join(kegg_genes_to_ko, by=c("kegg_entrez_id" = "kegg_entrez_id") )     %>%
      distinct() %>%
      group_by(Query_Species, Query_Strain, query_acc.ver, kegg_pathway_id, cleaned_kegg_pathway_id, kegg_ko_id ) %>%
      # partition(Query_Species, Query_Strain, query_acc.ver, kegg_pathway_id, kegg_ko_id, .cluster = cluster) %>%
      summarise( evalue = min(evalue)) %>%
      # collect() %>%
      ungroup() %>%
      inner_join( chromosome_id_to_protein_id, by=c("query_acc.ver" = "protein_refseq_id") ) %>%
      inner_join ( strain_to_accessions_cleaned, by =c("chromosome_refseq_id" = "chromosome_refseq_id", 
                                                       "Query_Species" = "Species",
                                                       "Query_Strain" = "Strain"))

  return( hits_to_pathways_table_cleaned )
}


# hits_to_pathways_table_cleaned <- read_tsv( file=file.path(kegg_annotations_dir, "hits_to_pathways_uniprot_mapping.tsv") ) 

```

## Summary statistics of how many protein hits are mapped to KEGG pathways
```{r}

summary_statistics_hits_to_KEGG_pathways <-  function( hits_to_pathways_table_cleaned ) {
    hits_to_pathways_uniprot_summary_statistics <- hits_to_pathways_table_cleaned  %>%
      distinct(Query_Species, Query_Strain, chromosome_refseq_id, query_acc.ver ) %>%
      group_by(Query_Species, Query_Strain, chromosome_refseq_id) %>%
      # partition(Query_Species, Query_Strain, chromosome_refseq_id, .cluster = cluster) %>%
      summarise( num_proteins_in_pathways = n()) %>%
      # collect() %>%
      ungroup() %>%
      inner_join(  num_seq_per_accession, by=c("chromosome_refseq_id" = "chromosome_refseq_id")   ) %>%
      dplyr::rename( total_num_proteins_in_proteome = "num_sequences") %>%
      distinct() 
    
    return( hits_to_pathways_uniprot_summary_statistics)
    
}

```

## Create function to loop through different e-values and then write the results to a different directory
```{r}

loop_evalues_map_hits_to_pathways_and_ko_groups <- function(my_hits_to_pathways_table, my_e_value_cutoff) {

  print(paste(my_e_value_cutoff, "\n"))
  
  kegg_annotations_dir <- file.path( oc_results_dir, "KEGG_Annotations", paste( "E_value_", my_e_value_cutoff, sep=""), "Map_to_Uniprot" )
    
  my_hits_to_pathways_table_filt <- my_hits_to_pathways_table %>%
                                    filter( evalue < my_e_value_cutoff  )
  
  hits_to_pathways_uniprot_summary_statistics <- summary_statistics_hits_to_KEGG_pathways(
                                                    my_hits_to_pathways_table_filt)

  write_tsv(hits_to_pathways_uniprot_summary_statistics, 
            path=file.path(kegg_annotations_dir, 
                           "mapping_uniprot_hits_to_pathways_summary_statistics.tsv") )
  # dim(my_hits_to_pathways_table)
  # dim(hits_to_pathway_ko_table)
  
  if( my_e_value_cutoff == best_e_value_to_use) {
       write_tsv(my_hits_to_pathways_table_filt, path=file.path(kegg_annotations_dir, "mapping_uniprot_hits_to_pathways_and_ko.tsv") )
  }

  return()
}
```


## Count how many query proteins are mapped to uniprot proteins
```{r}

gc()

partial_summary_uniprot <- purrr::partial( summary_statistics_hits_to_uniprot_save_file,
                                           hits_to_uniprot_table=hits_to_uniprot_table )

purrr::walk(list_of_e_values_to_test,  partial_summary_uniprot  )

gc()

```

## Cleaning the KEGG Database tables 
Lucid Chart is the program I've used to draw the entity-relationships diagram in the Google Doc file (https://www.lucidchart.com/documents/edit/bacc75e2-392e-420e-88d1-e62b75ebe939/0?callback=close&name=docs&callback_type=back&v=2025&s=592).

https://www.lucidchart.com/pages/ER-diagram-symbols-and-meaning

```{r}
kegg_genes_to_uniprot_cleaned  <- vroom::vroom(kegg_genes_to_uniprot_cleaned_file, 
                                               delim = "\t", 
                                               col_names = c("kegg_entrez_id", "uniprot_acc")) %>% 
          # Convert the uniprot_acc (e.g. up:P11245) to the uniprot_acc_cleaned (e.g. P11245)
          # by removing the prefix. 
					mutate( uniprot_acc_cleaned = str_replace_all( uniprot_acc, "^.*:", ""))

kegg_genes_to_pathway_cleaned  <- vroom::vroom(kegg_genes_to_pathway_cleaned_file, 
                                               delim = "\t", 
                                               col_names =  c("kegg_entrez_id", "kegg_pathway_id"))
kegg_genes_to_ko               <- vroom::vroom(kegg_genes_to_ko_file, 
                                               delim = "\t", 
                                               col_names = c("kegg_entrez_id", "kegg_ko_id"))
 

```

## Test all of the E-values in a purr::walk loop
```{r}
hits_to_pathways_table_cleaned <- convert_hits_to_pathways( hits_to_uniprot_table, 
                                                            max( list_of_e_values_to_test) )

rm(kegg_genes_to_uniprot_cleaned)
rm(kegg_genes_to_pathway_cleaned)
rm(kegg_genes_to_ko)

gc()

partial_loop_evalues  <- purrr::partial( loop_evalues_map_hits_to_pathways_and_ko_groups, 
                                         my_hits_to_pathways_table = hits_to_pathways_table_cleaned )

purrr::walk(list_of_e_values_to_test,  partial_loop_evalues)

gc()
```

## Generate the RefSeq ID to KEGG pathway ID lookup table
```{r}
kegg_annotations_my_evalue_dir <- file.path( oc_results_dir, 
                                             "KEGG_Annotations", 
                                             paste( "E_value_", 
                                                    best_e_value_to_use, 
                                                    sep=""), 
                                             "Map_to_Uniprot" )

my_hits_to_pathways_table_filt <- read_tsv( file=file.path(kegg_annotations_my_evalue_dir, 
                                  "mapping_uniprot_hits_to_pathways_and_ko.tsv") )

refseq_id_to_kegg_pathway_id_table <- my_hits_to_pathways_table_filt %>%
  distinct( query_acc.ver, Query_Species, Query_Strain, 
            cleaned_kegg_pathway_id) 
  

write_tsv(refseq_id_to_kegg_pathway_id_table, 
          path=file.path(kegg_annotations_my_evalue_dir,
                         "refseq_id_to_kegg_pathway_id_table.tsv") )
```









